### DISCUSSION TITLE
How do I avoid overflow when multiplying years?

### DISCUSSION THREAD

**Discussion by azzamsa** [BurntSushi/jiff#82] [Q&A] [answered]
Hi,

I wanted to port my Chrono code to jiff, but I got `attempt to multiply with overflow`.
The Chrono part works just fine.

```rust
pub fn hijri_to_julian_chrono(date: chrono::NaiveDate) -> i32 {
    ((((11 * date.year() + 3) / 30) as f32).floor()
        + ((354 * date.year()) as f32).floor()
        + ((30 * date.month() as u8) as f32).floor()
        - (((date.month() as u8 - 1) / 2) as f32).floor()
        + date.day() as f32
        + 1_948_440.0
        - 385.0) as i32
}

pub fn hijri_to_julian_jiff(date: jiff::civil::Date) -> i32 {
    ((((11 * date.year() + 3) / 30) as f32).floor()
        + ((354 * date.year()) as f32).floor()
        + ((30 * date.month() as u8) as f32).floor()
        - (((date.month() as u8 - 1) / 2) as f32).floor()
        + date.day() as f32
        + 1_948_440.0
        - 385.0) as i32
}

fn main() {
    assert_eq!(
        hijri_to_julian_chrono(chrono::NaiveDate::from_ymd_opt(1442, 8, 25).unwrap()),
        2459313
    );
    assert_eq!(
        hijri_to_julian_jiff(jiff::civil::date(1442, 8, 25)),
        2459313
    );
}
```

The original python code is [here](https://github.com/abougouffa/pyIslam/blob/eb5cd7af0f4ad2e63e5f07054de690c76454b3ba/pyIslam/baselib.py#L34)

**Comment by BurntSushi** [comment, marked as answer]
`date.year()` returns an `i16`. In this case, `1442`. Multiplying that by `354` gives `510468`, which overflows `i16`. Chrono's `date.year()` returns an `i32`, which doesn't result in an overflow here. So a simple fix is to convert to `i32` and then multiply: `((354 * i32::from(date.year()))`.

**Reply by azzamsa** [reply]
Would you be able to explain why jiff uses i16 for the year? 
What the migration would be at `i16:MAX`? 😀

**Reply by BurntSushi** [reply]
The minimum and maximum years supported by Jiff are `-9999` and `9999`, respectively. Those fit in an `i16`.
